### DISCUSSION TITLE
How can I use prisma to create a foreign key that does not have a unique index

### DISCUSSION THREAD

**Discussion by bigTeethLi** [prisma/prisma#12191] [Q&A] [answered]
On the official documents of Prisma, I only see how to use `@relation` to create a `unique index foreign key`, I want to create a non-unique index, this is not reflected in the official documents, how can I use Prisma to create a `non-unique index Foreign key`?
[Official document example](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/many-to-many-relations)
<img width="685" alt="1" src="https://user-images.githubusercontent.com/101170394/157188988-627d5ff6-6cc3-4b56-92da-0c0854e03794.png">

**Comment by austincrim** [comment, marked as answer]
Hey @bigTeethLi 👋 ,

This is actually a mistake in our docs. I just generated the migration using Postgres and Prisma 3.10 and the SQL looks like this:

```sql

-- CreateTable
CREATE TABLE "Post" (
    "id" SERIAL NOT NULL,
    "title" TEXT NOT NULL,

    CONSTRAINT "Post_pkey" PRIMARY KEY ("id")
);

-- CreateTable
CREATE TABLE "Category" (
    "id" SERIAL NOT NULL,
    "name" TEXT NOT NULL,

    CONSTRAINT "Category_pkey" PRIMARY KEY ("id")
);

-- CreateTable
CREATE TABLE "CategoriesOnPosts" (
    "postId" INTEGER NOT NULL,
    "categoryId" INTEGER NOT NULL,
    "assignedAt" TIMESTAMP(3) NOT NULL DEFAULT CURRENT_TIMESTAMP,
    "assignedBy" TEXT NOT NULL,

    CONSTRAINT "CategoriesOnPosts_pkey" PRIMARY KEY ("postId","categoryId")
);

-- AddForeignKey
ALTER TABLE "CategoriesOnPosts" ADD CONSTRAINT "CategoriesOnPosts_postId_fkey" FOREIGN KEY ("postId") REFERENCES "Post"("id") ON DELETE RESTRICT ON UPDATE CASCADE;

-- AddForeignKey
ALTER TABLE "CategoriesOnPosts" ADD CONSTRAINT "CategoriesOnPosts_categoryId_fkey" FOREIGN KEY ("categoryId") REFERENCES "Category"("id") ON DELETE RESTRICT ON UPDATE CASCADE;

```

Per my colleague, Prisma only generates a unique foreign key index in two cases:
- when using `@unique` or `@@unique` 
- a one-to-one relation (one side required, one side optional, no list fields). We use a unique constraint there to make sure it's a 1:1 relation and does not accidentally become one-to-many. There is no corresponding @(@)unique in the schema.

Thanks for catching this, I will make sure it gets fixed!

**Reply by bigTeethLi** [reply]
Hey @austincrim 👋
Thank you for your detailed answer, through which I located my problem. it's really helpful to me!
And I found a phenomenon that I'm not sure if it's normal:
If I initially create a 1:1 relation and then need to change it to a many-to-many relation, the foreign key index does not change from unique to non-unique. And need to drop the table and then re-create, the foreign key index can be updated.

Using the above example: 
Create 1:1 relation first:
```prisma
model Post {
  id         Int                 @id @default(autoincrement())
  title      String
  categories CategoriesOnPosts?
}

model Category {
  id    Int                 @id @default(autoincrement())
  name  String
  posts CategoriesOnPosts?
}

model CategoriesOnPosts {
  post       Post     @relation(fields: [postId], references: [id])
  postId     Int // relation scalar field (used in the `@relation` attribute above)
  category   Category @relation(fields: [categoryId], references: [id])
  categoryId Int // relation scalar field (used in the `@relation` attribute above)
  assignedAt DateTime @default(now())
  assignedBy String

  @@id([postId, categoryId])
}
```
Run `prisma generate && prisma db push`

And then change the relation to Many-to-Many(only "?" to "[]" )
```prisma
model Post {
  id         Int                 @id @default(autoincrement())
  title      String
  categories CategoriesOnPosts[]
}

model Category {
  id    Int                 @id @default(autoincrement())
  name  String
  posts CategoriesOnPosts[]
}

model CategoriesOnPosts {
  post       Post     @relation(fields: [postId], references: [id])
  postId     Int // relation scalar field (used in the `@relation` attribute above)
  category   Category @relation(fields: [categoryId], references: [id])
  categoryId Int // relation scalar field (used in the `@relation` attribute above)
  assignedAt DateTime @default(now())
  assignedBy String

  @@id([postId, categoryId])
}
```
Run `prisma generate && prisma db push`.

But the foreign key is still a unique index:
<img width="720" alt="image" src="https://user-images.githubusercontent.com/101170394/157794781-89fee53d-3aff-4e34-a49b-14855c000130.png">
If I drop the `CategoriesOnPosts` table and run `prisma generate && prisma db push`, the foreign key can only be changed to non-unique indexes
<img width="707" alt="image" src="https://user-images.githubusercontent.com/101170394/157796436-a9cacf29-10ef-4aa1-bd07-ad991f90bf3e.png">


If the many-to-many relation is changed to a 1:1 relation, the indexes of the foreign key can be updated directly.

**Reply by austincrim** [reply]
That looks unexpected! Would you mind creating a separate [bug report](https://github.com/prisma/prisma/issues/new?assignees=&labels=kind%2Fbug&template=bug_report.yml) with this info? Our engineering teams will pick it up from there.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
